File: app/src/main/java/com/agotakiss/androidtest/injector/Injector.java
package com.agotakiss.androidtest.injector;

import com.agotakiss.androidtest.data.network.MovieDbApi;
import com.agotakiss.androidtest.data.network.RetrofitHelper;
import com.agotakiss.androidtest.data.repository.GenreRepositoryImpl;
import com.agotakiss.androidtest.data.repository.MovieRepositoryImpl;
import com.agotakiss.androidtest.data.store.GenreStore;
import com.agotakiss.androidtest.data.store.GenreStoreImpl;
import com.agotakiss.androidtest.domain.repository.GenreRepository;
import com.agotakiss.androidtest.domain.repository.MovieRepository;

import retrofit2.Retrofit;

public class Injector {

    private static Retrofit retrofit;

    public static Retrofit getRetrofit(){
        if(retrofit == null){
            retrofit = RetrofitHelper.initRetrofit();
        }
        return retrofit;
    }

    private static MovieDbApi movieDbApi;

    public static MovieDbApi getMovieDbApi(){
        if(movieDbApi == null){
            movieDbApi = getRetrofit().create(MovieDbApi.class);
        }
        return movieDbApi;
    }

    private static GenreRepository genreRepository;

    public static GenreRepository getGenreRepository(){
        if(genreRepository == null){
            genreRepository = new GenreRepositoryImpl();
        }
        return genreRepository;
    }

    private static MovieRepository movieRepository;

    public static MovieRepository getMovieRepository(){
        if(movieRepository == null){
            movieRepository = new MovieRepositoryImpl();
        }
        return movieRepository;
    }

    private static GenreStore genreStore;

    public static GenreStore getGenreStore(){
        if(genreStore == null){
            genreStore = new GenreStoreImpl();
        }
        return genreStore;
    }


}

File: app/src/main/java/com/agotakiss/androidtest/di/Injector.kt
package com.agotakiss.androidtest.di

import com.agotakiss.androidtest.data.network.MovieDbApi
import com.agotakiss.androidtest.data.network.RetrofitHelper
import com.agotakiss.androidtest.data.repository.CastRepositoryImpl
import com.agotakiss.androidtest.data.repository.GenreRepositoryImpl
import com.agotakiss.androidtest.data.repository.MovieRepositoryImpl
import com.agotakiss.androidtest.data.store.GenreStore
import com.agotakiss.androidtest.data.store.GenreStoreImpl
import com.agotakiss.androidtest.domain.repository.CastRepository
import com.agotakiss.androidtest.domain.repository.GenreRepository
import com.agotakiss.androidtest.domain.repository.MovieRepository
import retrofit2.Retrofit


object Injector {

    private var retrofit: Retrofit? = null

    private var movieDbApi: MovieDbApi? = null

    private var genreRepository: GenreRepository? = null

    private var movieRepository: MovieRepository? = null

    private var castRepository : CastRepository? = null

    private var genreStore: GenreStore? = null

    fun getRetrofit(): Retrofit {
        if (retrofit == null) {
            retrofit = RetrofitHelper.initRetrofit()
        }
        return retrofit!!
    }

    fun getMovieDbApi(): MovieDbApi {
        if (movieDbApi == null) {
            movieDbApi = getRetrofit().create(MovieDbApi::class.java)
        }
        return movieDbApi!!
    }

    fun getGenreRepository(): GenreRepository {
        if (genreRepository == null) {
            genreRepository = GenreRepositoryImpl()
        }
        return genreRepository!!
    }

    fun getMovieRepository(): MovieRepository {
        if (movieRepository == null) {
            movieRepository = MovieRepositoryImpl()
        }
        return movieRepository!!
    }
    fun getCastRepository(): CastRepository {
        if (castRepository == null) {
            castRepository = CastRepositoryImpl()
        }
        return castRepository!!
    }

    fun getGenreStore(): GenreStore {
        if (genreStore == null) {
            genreStore = GenreStoreImpl()
        }
        return genreStore!!
    }
}
